[Woo POS] Loading state when pull-to-refresh shouldn't be dashboard-wide

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of https://github.com/woocommerce/woocommerce-ios/issues/14186
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
Following the separation between first data fetch vs subsequent data fetch, this PR separates the dashboard loading state more granularly: Initial loading and loading. This makes an easier separation between the initial loading screen when we first enter the POS mode versus other loading states that the view may be in (example: pull-to-refresh).

An immediate benefit is that fixes a current bug where the POS initial loading screen was shown each time we pull-to-refresh.

I've tried to simplify the `ItemListState` a bit further by removing `.empty` in favour of using `.loaded([])` but there's a few hiccups when we observe item list state and requires further refactoring, so I left that bit outside of this PR.


**Before:**

https://github.com/user-attachments/assets/a9ce00ec-77fc-4ca9-b689-e286a234d524

**After:**

https://github.com/user-attachments/assets/32b0a0d3-5200-4ec7-b61d-03247ce7fdfe

## Testing
- In POS, observe how pull-to-refresh does not display the full-width loading view anymore, but its encapsulated to the items list view
---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
Add new initialLoading state

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Merge branch 'trunk' into task/pos-pagination-update-loading-state-ptr

iamgabrielma committed:
Update existing tests

iamgabrielma committed:
Add missing equatable conformance

Tests were failing for equality when comparing loading state, as one of the cases was missing

jaclync commented on WooCommerce/WooCommerceTests/POS/ViewModels/PointOfSaleDashboardViewModelTests.swift:
```diff
@@ -297,16 +297,7 @@ final class PointOfSaleDashboardViewModelTests: XCTestCase {
-    func test_isInitialLoading_when_item_list_loading_and_empty_then_true() {
-        // Given
-        mockItemListViewModel.items = []
-        mockItemListViewModel.state = .loading
```
non-blocking ❓ : is it worth keeping this test case with this state being `.initialLoading`?

iamgabrielma replied:
It doesn't hurt! Restored on https://github.com/woocommerce/woocommerce-ios/pull/14226/commits/19dac743b7d68a119541c51431888f85954dd67f

jaclync commented on WooCommerce/WooCommerceTests/POS/ViewModels/ItemListViewModelTests.swift:
```diff
@@ -407,7 +407,66 @@ final class ItemListViewModelTests: XCTestCase {
+        itemProvider.shouldReturnZeroItems = true
+        let sut = ItemListViewModel(itemProvider: itemProvider)
+
+        let expectation = XCTestExpectation(description: "Publisher should emit state changes")
```
❓ My understanding is that an XCTest expectation has to be waited for by `wait(for:)` to take effect, which doesn't seem needed in these two new test cases since the load functions are already `async` 🤔 I tried removing the expectation in this and the next test case and tests seemed to pass locally. Do you plan to wait for the expectation?

iamgabrielma replied:
That makes a lot of sense 😅 we do not need to wait here and this can be removed: https://github.com/woocommerce/woocommerce-ios/pull/14226/commits/c07640129655f7fc364bc82e4ae1dd57f37a95a7

jaclync reviewed (approved):
LGTM! One non-blocking observation: during pull-to-refresh, the product list view along with its scroll indicator seems to move around, and the spinner is overlapping with the fading non-simple products notice briefly. Feel free to consider improvements here later.


https://github.com/user-attachments/assets/e3354e15-2e8c-4897-a738-b079799a64f6

iamgabrielma committed:
restore unit test

iamgabrielma committed:
remove unnecessary XCTestExpectation

iamgabrielma commented:
Thanks for the review! I've addressed the comments except the following, which I [logged separately](https://github.com/woocommerce/woocommerce-ios/issues/14251):

>  One non-blocking observation: during pull-to-refresh, the product list view along with its scroll indicator seems to move around, and the spinner is overlapping with the fading non-simple products notice briefly. Feel free to consider improvements here later.

Since we might remove/update the banner (and most likely rework animations) with the latest design update I'll leave this one as a low-priority for the moment.
